cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+  included file COSL.txt.
+*/
+
+#include <package_module.h>
+#include <pipes.h>
+#include <signals.h>
+#include <buffer.h>
+#include <string_lib.h>
+#include <actuator.h>
+#include <file_lib.h>
+#include <known_dirs.h>
+#include <locks.h>
+#include <rlist.h>
+#include <policy.h>
+#include <eval_context.h>
+
+#define INVENTORY_LIST_BUFFER_SIZE 100 * 80 /* 100 entries with 80 characters 
+                                             * per line */
+
+static bool UpdateSinglePackageModuleCache(EvalContext *ctx,
+                                    const PackageModuleWrapper *module_wrapper,
+                                    UpdateType type, bool force_update);
+static char *GetPackageModuleRealPath(const char *package_manager_name);
+static int NegotiateSupportedAPIVersion(PackageModuleWrapper *wrapper);
+
+
+void DeletePackageModuleWrapper(PackageModuleWrapper *wrapper)
+{
+    free(wrapper->path);
+    free(wrapper->name);
+    free(wrapper);
+}
+
+PackageModuleWrapper *NewPackageModuleWrapper(PackageModuleBody *package_module)
+{
+    assert(package_module && package_module->name);
+
+    PackageModuleWrapper *wrapper = xmalloc(sizeof(PackageModuleWrapper));
+    
+    wrapper->path = GetPackageModuleRealPath(package_module->name);
+    wrapper->name = SafeStringDuplicate(package_module->name);
+    wrapper->package_module = package_module;
+    
+    /* Check if file exists */
+    struct stat sb;
+    if (!wrapper->path || (stat(wrapper->path, &sb) != 0))
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "can not find package wrapper in provided location: %s",
+            wrapper->path);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    
+    /* Negotiate API version */
+    wrapper->supported_api_version = NegotiateSupportedAPIVersion(wrapper);
+    if (wrapper->supported_api_version != 1)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "unsupported package module wrapper API version: %d",
+            wrapper->supported_api_version);
+        DeletePackageModuleWrapper(wrapper);
+        return NULL;
+    }
+    return wrapper;
+}
+
+
+void UpdatePackagesCache(EvalContext *ctx, bool force_update)
+{
+    Log(LOG_LEVEL_DEBUG, "Updating package cache.");
+    
+    PackagePromiseGlobalLock package_lock = 
+            AcquireGlobalPackagePromiseLock(ctx);
+                        
+    if (package_lock.g_lock.lock == NULL)
+    {
+        Log(LOG_LEVEL_INFO, "Can not acquire global lock for package promise. "
+            "Skipping updating cache.");
+        return;
+    }
+                        
+    Rlist *default_inventory = GetDefaultInventoryFromContext(ctx);
+    
+    for (const Rlist *rp = default_inventory; rp != NULL; rp = rp->next)
+    {
+        const char *pm_name =  RlistScalarValue(rp);
+        
+        /* We don't want inventory to be reported. */
+        if (StringSafeEqual(pm_name, "cf_null"))
+        {
+            break;
```

Comment:
We can either have some "valid" entry on the list of it can be defined as empty list in the policy. If this is an empty list there is no need to continue iteration.